File: src/draft/draft20241018/494.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var findTargetSumWays = function (nums, target) {
  /**
   * 整个nums的和假设为sum
   * 选择正数的和为 p  选择负数的和的绝对值为 q
   * p + q = sum
   * p - q = target
   * p = (sum + target)/2   上述两式相加 
   * q = (sum - target)/2   上述两式相减
   */
  let realTarget = (nums.reduce((pre, cur) => pre + cur) + Math.abs(target))
  // 不能得到合法的
  if (realTarget < 0 || realTarget % 2) {
    return 0
  }
  realTarget = realTarget / 2
  const n = nums.length
  let memo = Array.from({ length: n }, () => new Array(realTarget + 1).fill(-1))
  // i是第几个位置， c是容量
  const dfs = (i, c) => {
    // 如果到边界了 刚好c没有容量了则表示是一个可行方案
    if (i < 0) return c == 0 ? 1 : 0
    if (memo[i][c] != -1) return memo[i][c]
    if (c < nums[i]) {
      // 只能不选
      return dfs(i - 1, c)
    }
    return dfs(i - 1, c) + dfs(i - 1, c - nums[i])
  }
  return dfs(n - 1, realTarget)
};

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var findTargetSumWays = function (nums, target) {
  /**
   * 整个nums的和假设为sum
   * 选择正数的和为 p  选择负数的和的绝对值为 q
   * p + q = sum
   * p - q = target
   * p = (sum + target)/2   上述两式相加 
   * q = (sum - target)/2   上述两式相减
   */
  let realTarget = (nums.reduce((pre, cur) => pre + cur) + Math.abs(target))
  // 不能得到合法的
  if (realTarget < 0 || realTarget % 2) {
    return 0
  }
  realTarget = realTarget / 2
  const n = nums.length
  // dfs中有i-1  所以需要n+1
  let dp = Array.from({ length: n + 1 }, () => new Array(realTarget + 1).fill(0))
  dp[0][0] = 1
  for (let i = 0; i < n; i++) {
    for (let c = 0; c <= realTarget; c++) {
      if (c < nums[i]) {
        dp[i + 1][c] = dp[i][c]
      } else {
        dp[i + 1][c] = dp[i][c] + dp[i][c - nums[i]]
      }
    }
  }
  return dp[n][realTarget]
};
// 空间优化的写法
var findTargetSumWays = function (nums, target) {
  /**
   * 整个nums的和假设为sum
   * 选择正数的和为 p  选择负数的和的绝对值为 q
   * p + q = sum
   * p - q = target
   * p = (sum + target)/2   上述两式相加 
   * q = (sum - target)/2   上述两式相减
   */
  let realTarget = (nums.reduce((pre, cur) => pre + cur) + Math.abs(target))
  // 不能得到合法的
  if (realTarget < 0 || realTarget % 2) {
    return 0
  }
  realTarget = realTarget / 2
  const n = nums.length
  let dp = Array.from({ length: 2 }, () => new Array(realTarget + 1).fill(0))
  dp[0][0] = 1
  for (let i = 0; i < n; i++) {
    for (let c = 0; c <= realTarget; c++) {
      if (c < nums[i]) {
        dp[(i + 1) % 2][c] = dp[i % 2][c]
      } else {
        dp[(i + 1) % 2][c] = dp[i % 2][c] + dp[i % 2][c - nums[i]]
      }
    }
  }
  return dp[n % 2][realTarget]
};

File: src/module/classify/18/494.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        s = sum(nums) - abs(target)
        if s < 0 or s % 2:
            return 0
        m = s // 2  # 背包容量

        @cache  # 缓存装饰器，避免重复计算 dfs 的结果（记忆化）
        def dfs(i: int, c: int) -> int:
            if i < 0:
                return 1 if c == 0 else 0
            if c < nums[i]:
                return dfs(i - 1, c)  # 只能不选
            return dfs(i - 1, c) + dfs(i - 1, c - nums[i])  # 不选 + 选
        return dfs(len(nums) - 1, m)



class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        s = sum(nums) - abs(target)
        if s < 0 or s % 2:
            return 0
        m = s // 2  # 背包容量

        n = len(nums)
        f = [[0] * (m + 1) for _ in range(n + 1)]
        f[0][0] = 1
        for i, x in enumerate(nums):
            for c in range(m + 1):
                if c < x:
                    f[i + 1][c] = f[i][c]  # 只能不选
                else:
                    f[i + 1][c] = f[i][c] + f[i][c - x]  # 不选 + 选
        return f[n][m]




class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        s = sum(nums) - abs(target)
        if s < 0 or s % 2:
            return 0
        m = s // 2  # 背包容量

        n = len(nums)
        f = [[0] * (m + 1) for _ in range(2)]
        f[0][0] = 1
        for i, x in enumerate(nums):
            for c in range(m + 1):
                if c < x:
                    f[(i + 1) % 2][c] = f[i % 2][c]  # 只能不选
                else:
                    f[(i + 1) % 2][c] = f[i % 2][c] + f[i % 2][c - x]  # 不选 + 选
        return f[n % 2][m]



class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        s = sum(nums) - abs(target)
        if s < 0 or s % 2:
            return 0
        m = s // 2  # 背包容量

        f = [1] + [0] * m
        for x in nums:
            for c in range(m, x - 1, -1):
                f[c] += f[c - x]
        return f[m]
